### ISSUE TITLE
Extract files from VPK?

### ISSUE THREAD

**Issue by lf94** [snake-biscuits/bsp_tool#7]
I'm currently trying to interact with the VPK using https://github.com/ValvePython/vpk but in my two samples, it says it's missing a magic number, and in the other it says it's invalid.

Does Apex Legends use some special VPK variant?

**Comment by lf94** [comment]
From some quick research here on GitHub it seems there are two other projects which attempt to write a VPK extractor:

* https://github.com/r2d2rigo/ApexVpkInventory
* https://github.com/Unordinal/UnoVPKTool

Inspecting the vpk with `hexdump` leads me to believe it's a variant...

**Comment by lf94** [comment]
After some more searching, it appears https://dev.cra0kalo.com/ has worked on a closed source VPK extractor over the years. That really sucks. Looks like I'm going to have to write one based on the available knowledge.

**Comment by lf94** [comment]
https://wiki.modme.co/wiki/apps/Legion.html seems to be the leader in this space.

**Comment by lf94** [comment]
englishclient_mp_rr_canyonlands_64k_x_64k.bsp.pak000_dir.vpk seems to be a proper vpk file with file listings inside.

**Comment by snake-biscuits** [comment]
the `pak000_dir.vpk`s host all headers which point to files, cra0kalo has said they were going to release their source code for opening .vpks but they haven't yet

IIRC you can still get cra0's tool just to extract the files, I've extracted most of Apex's files into a folder on my pc and I just load them with bsp_tool when I want to check something

cra0 explained the .vpk header here: https://forum.xentax.com/viewtopic.php?f=10&t=11208
but I think it's changed since 2014

**Comment by lf94** [comment]
@snake-biscuits yep, that's what I observed last night. Really I think there's no reason to wait on cra0kalo - the format looks pretty straight forward. The link is for sure outdated; the vpk's no longer have that checksum information for example. From what I can tell they are very close to a vpk v1 "header".

Could you please send me cra0kalo's tool at inbox@leefallat.ca ? I would greatly appreciate it! :slightly_smiling_face: 

Tonight I will probably go ahead and use the tool on my Windows machine (doing all the current work in Debian with r2cutter and bless) to start getting familiar with what's actually inside the vpk.

**Comment by lf94** [comment]
@snake-biscuits I also had the thought it's very possible the collision mesh is stored in rpak files, since it appears they use rpaks for more modular updates, and I can imagine from a development point of view that collision needs to be updated way more often than graphical mesh.

**Comment by lf94** [comment]
Using https://zenhax.com/viewtopic.php?t=3461 I should be able to reverse the rpak format. The dll is not mangled like every other single thing I've looked at so far :slightly_smiling_face: The decompression routine is not very large either.

**Comment by snake-biscuits** [comment]
You might find some of the tools & modding guides here helpful: https://noskill.gitbook.io/titanfall2/

**Comment by lf94** [comment]
Thanks :slightly_smiling_face: 

So I've made some progress on RPak, but not much:

```
struct RPakHeader {
  magic: u32,
  version: u16,
  section: u16, // some switch between where to start
}
```

I found the actual routine which decompresses: `180004b80`. Unfortunately there is still some stuff outside of it which I think I need to understand. I've spent most of my time trying to figure out how the section property is used and how the expected decompressed size is calculated.

**Comment by snake-biscuits** [comment]
https://user-images.githubusercontent.com/36507175/108614855-eb0f1180-7449-11eb-8833-3d6586c33f68.mp4

Someone recently sent me this, it's the apex firing range with an edited vertex lump
They seem to hover in place until the cut the UDP stream part of their server connection
Seems like the physics is at least based on those vertices & the server side simulates physics and those physics override clientside player physics

**Comment by lf94** [comment]
@snake-biscuits sorry for the absence - life has been busy.

That video is interesting but does it really tell us anything we don't already know? Of course physics / collision must be checked server-side, otherwise it would be very easy for players to walk through walls and other cheats.

I've suspected that the game takes the defined vertices, and creates some collision planes out of them, as a way to save space. Has the person tried walking into the new mesh they've defined? That would be more interesting, and I suspect that they will walk right through it.

**Comment by snake-biscuits** [comment]
I have found that the angel city vpk contains some different versions of the map in depot/r2pcprecert/game/r2/maps. This version has some slight differences (vertex lump included) but they’re pretty small changes

From a glance it seems mostly to be changes to some optimisation triggers but those differences could be handy for identifying some format stuff

**Comment by lf94** [comment]
Hey @snake-biscuits any progress on this great mystery? :stuck_out_tongue:

**Comment by snake-biscuits** [comment]
I’ve started work on a blender addon for viewing bsps, so far it’s been pretty helpful for understanding how everything fits together.
I think can start working towards looking at loading stuff like the planes lump to see how it fits in
Maybe even edit that lump to see if it affects physics in-game

I am pretty busy with uni at the moment but I will still be working on that.
Did find that both titanfalls & apex have a mp_lobby map in the common .vpk, that’ll be pretty useful as it’s the smallest map in each game, though apex’s is pretty complex

**Comment by snake-biscuits** [comment]
![1C88E8F5-3036-4D86-BF2B-9B5E59D4FCEE](https://user-images.githubusercontent.com/36507175/116526063-8924c980-a91c-11eb-8ece-5f293de074fa.png)
![D2E92BD3-E53F-4671-8AAB-597C0248359E](https://user-images.githubusercontent.com/36507175/116526085-8f1aaa80-a91c-11eb-9d95-8232f6975d02.png)
![9B5AC15F-8597-44A9-939B-5FD568246F76](https://user-images.githubusercontent.com/36507175/116526090-8fb34100-a91c-11eb-9c33-9587bffc2ab6.png)

**Comment by lf94** [comment]
This is excellent work @snake-biscuits ! I really wish I could do more but as we know reversing work can be so time consuming. What's that second screenshot? :o

**Comment by snake-biscuits** [comment]
I've made some progress on understanding physics in titanfall 1 & 2 this past week.
having reverse engineered about half of the `PHYSICS_COLLIDE` lump, I noticed Titanfall 2 barely uses the lump
All Titanfall 2's `PHYSICS_COLLIDE` lumps only use the lump for fog volumes as far as I can see.
However Titanfall 1 uses the lump just like a Source Engine .bsp

Looking at this GDC talk:
https://www.youtube.com/watch?v=6BIfqfC1i7U
I think they started moving towards using the Meshes / Tricoll lumps

Since Apex cuts out a lot of the `CM_*` (Collision Model) lumps, I think Apex uses a combination of Meshes & the `TRICOLL_*` lumps to handle physics.

Will let you know more as I investigate

**Comment by lf94** [comment]
Excellent work as always @snake-biscuits , I am still following you and always excited to see a GitHub email from you pop up.

Could you link to any material on TRICOLL? Does it stand for triangle collision?

**Comment by snake-biscuits** [comment]
Recently we've been digging around in the [Titanfall | 2: Northstar Discord Server](https://discord.gg/34prExHE) to work out how collision meshes are stored in Titanfall 2 maps.

So far we've worked out prop physics is loaded from the props at runtime and the PHYSICS_COLLIDE lump doesn't hold as much as we think

In attempting to port some titanfall 1 maps to titanfall 2 we've encountered some ghost collidables (where props should be, but we disable props to make the map run) and some crashes that appear to happen as the result of colliding with *something*

One thing we have noticed is that collisions with visible geometry appears to be directly based on the position of that renderable geometry

Invisible walls (clipping brushes etc.), we haven't yet identified

**Comment by snake-biscuits** [comment]
Added [`apex_legends.BVHNode`](https://github.com/snake-biscuits/bsp_tool/blob/master/bsp_tool/branches/respawn/apex_legends.py#L298), which might be pivotal to how physics works in Apex
Still got a bunch of new structs added by @r-ex in [`src/structs_r1.h`](https://github.com/snake-biscuits/bsp_tool/blob/master/src/structs_r1.h) to port over to python so we can up the coverage of the Titanfall Engine & start to better map out how everything connects

In other news [`MRVN-radiant`](https://github.com/MRVN-radiant/MRVN-radiant) is starting to get some serious use and is making for a really good testbed for building test maps to find out what works and what doesn't

Mostly brute forcing but it's still something vs. just statistical analysis

Making a lot of progress in general atm, tho mostly focused on Titanfall 2 maps rn

**Comment by lf94** [comment]
Amazing :) I'm still here watching from the sidelines!

Has anyone tried putting the PC versions of Apex into IDA Pro or Ghidra and find the routines for collision?

**Comment by snake-biscuits** [comment]
yeah we just did w/ some r5reloaded devs, haven't figured out how it all fits together yet but we can stop in the middle of a lookup

**Comment by r-ex** [comment]
https://www.youtube.com/watch?v=1WgViRE7rmg result of returning in some collision testing function while in game

**Comment by snake-biscuits** [comment]
[MRVN-Radiant](https://github.com/MRVN-Radiant/MRVN-Radiant) can now make basic maps for Apex Legends w/ physics
https://github.com/MRVN-Radiant/MRVN-Radiant/blob/3b89261ccb37f9baa1f0561f38db23c4cdc91035/tools/remap/source/bspfile_apex_legends.cpp#L268-L290

**Comment by snake-biscuits** [comment]
Haven't looked at exporting / converting to mesh of any kind just yet tho.
I've yet to dig into the Apex Legends side much myself, my focus has mostly been on Titanfall 2 lately

**Comment by F1F7Y** [comment]
> MRVN-Radiant can now make basic maps for Apex Legends w/ physics

It can't. I just wanted to clean up the apex functions a bit so I decided to create those functions even though they dont do anything.

There's some interesting information I could share though. My current theory is this:
The game walks a [BVH4 tree](https://github.com/MRVN-Radiant/MRVN-Radiant/blob/3b89261ccb37f9baa1f0561f38db23c4cdc91035/tools/remap/source/bspfile_apex_legends.h#L88-L105) to find out which primitives are close to the player. The nodes themselves index into a collision mask which determines who can collide with the node and its child nodes ( titan only, player only, etc. Amos pulled out the flags from apex [here](https://github.com/Mauler125/r5sdk/blob/437b80eac4f039c56863f3cef07baeba187ea325/r5dev/public/bspflags.h#L24-L93). I also confirmed most of them in titanfall ).

The nodes also store a child type, this can be 1 for no child, 0 for bvh node, and then 14 unknown types. From bounds checking I confirmed most of the unknown types index into a separate lump called BVH Leaf Data. The game checks if the size of the lump is divisible by 4 which suggests it's 4 bytes. Thing is I havent been able to spot a pattern and arent good enough in ida/ghidra(which just doesnt want to install) to look into the executable. I'd expect BVH Leaf Data to then index Texture Data and vertices.

One last thing regarding my collision stub. The game crashes when there's only one BVH Node with no children. It needs to index at least once into BVH Leaf Data. This makes sense as you kinda need collision to play a map.

**Comment by F1F7Y** [comment]
So theoratically we need to reverse the bounds information in BVHNodes and figure out how BVHLeafData is structured and indexed.

**Comment by lf94** [comment]
What are "primitives" defined as here?

It's been awhile: does all this mean the geometry mesh is doubly used as a collision mesh? How does the game stop players from jumping past an invisible fence? Is it just transparent geometry?

**Comment by F1F7Y** [comment]
> What are "primitives" defined as here?

Titanfall uses brushes ( a convex volume defined by planes ) and tricoll ( a mesh made of vertices ) for collision. Respawn refers to these as collision primitives. We currently only generate brushes as we dont have all structs for tricoll reversed.

So in this context I'd expect each BVHNode type to define a primitive type which is then read from BVHLeafData. From looking at maps it seems respawn only use the first ~12 types.

> It's been awhile: does all this mean the geometry mesh is doubly used as a collision mesh? How does the game stop players from jumping past an invisible fence? Is it just transparent geometry?

I'd expect BVHLeafData to define it's own meshes as the ones used for visuals can be quite large which is why they'd be inefficient for collision testing.

**Comment by r-ex** [comment]
here's a couple of videos showing the (rather chaotic) visualisation of some of the BVH4 leaf data that is used for testing collision around the map:
https://www.youtube.com/watch?v=APrxrT18EOs (triangles)
https://www.youtube.com/watch?v=l80fKAsdUQM (quads)

**Comment by lf94** [comment]
It appears that some sort of basic bounding volume algorithm is used for some things like rock faces. I think this pretty much explains that they generate collision from graphical meshes and then probably tweak it after...

**Comment by snake-biscuits** [comment]
That makes sense, iirc Apex Legends splits VIS with static props, so they need some kind of inner volume for that
(for the big rock faces anyway)

Haven't looked enough at Apex' `.rmdl` to say much about how the physics model work tho, but afaik it's similar to the BVH4 system in `.bsp`

**Comment by r-ex** [comment]
> Haven't looked enough at Apex' `.rmdl` to say much about how the physics model work tho, but afaik it's similar to the BVH4 system in `.bsp`

yea, it's p much identical (same nodes, same type of leaf data) apart from some slight differences in how data is located (bsp uses MODELS lump, rmdl has a set of offsets in a sort of "header" struct)

one notable difference that @IJARika noticed in rmdl is that the most commonly used leaf primitive types (as indicated in the node data) were different from those used in BSP.
RMDL mostly used type 5 with a couple of models using 7 (haven't confirmed what either of these are yet, though i would bet that it's tris/quads respectively)
BSP used a mixture of type 4 (confirmed to be tris) and type 6 (confirmed to be quads)

**Comment by snake-biscuits** [comment]
Did a little research in #111 bringing bsp_tool a little more up to speed w/ rexx & rika’s work

haven’t gotten too deep yet, as rn I’m working on saving edits
once that’s working I can experiment in-game
though I want to check out some stuff with titanfall physics first (porting tf1->2)
